packages/venia-ui/lib/components/CheckoutPage/PaymentInformation/brainTreeDropIn.js

```diff
@@ -0,0 +1,173 @@
+ * 2) On submission (triggered by a parent), request the payment nonce.
+ */
+const BraintreeDropin = props => {
+    const {
+        onError,
+        onReady,
+        onSuccess,
+        shouldRequestPaymentNonce,
+        containerId,
+        shouldTeardownDropin,
+        resetShouldTeardownDropin
+    } = props;
+    const classes = mergeClasses(defaultClasses, props.classes);
+    const [isError, setIsError] = useState(false);
+    const [dropinInstance, setDropinInstance] = useState();
+
+    const createDropinInstance = useCallback(
+        async didClose => {
+            try {
+                const {
+                    default: dropIn
+                } = await import('braintree-web-drop-in');
+                const dropinInstance = await dropIn.create({
+                    authorization,
+                    container: `#${containerId}`,
+                    card: {
+                        cardholderName: {
+                            required: true
+                        },
+                        overrides: {
+                            fields: {
+                                number: {
+                                    maskInput: {
+                                        // Only show last four digits on blur.
+                                        showLastFour: true
+                                    }
+                                }
+                            }
+                        }
+                    }
+                });
+
+                if (didClose) {
+                    // If we get here but the form is closed we should teardown
+                    // instead of trying to use the instance.
+                    dropinInstance.teardown();
+                } else {
+                    setDropinInstance(dropinInstance);
+                    onReady(true);
+                }
+            } catch (err) {
+                console.error(
+                    `Unable to initialize Credit Card form (Braintree). \n${err}`
+                );
+
+                if (!didClose) {
+                    // Error state used to render error view. If we got an error
+                    // but we are closing the form we shouldn't try to update
+                    // state.
+                    setIsError(true);
+                }
+            }
+        },
+        [onReady, containerId]
+    );
+
+    useEffect(() => {
+        let didClose = false;
+
+        // Initialize the dropin with a reference to the container mounted in
+        // this component. We do this via onmount because we have to be sure
+        // the container id has mounted per braintree's API.
+        createDropinInstance(didClose);
+
+        // If we close the form quickly after opening we may end up in a
+        // semi-mounted state so we need a local variable to make sure we
+        // behave correctly.
+        return () => {
+            didClose = true;
+        };
```

Comment:
Now that I look a little closer at this effect I realize it is not functioning as intended with `didClose`. The original effect from the dropin file defined `didClose` within the same closure as the definition for the `createDropinInstance` function. This meant that when the component unmounted and the effect's callback was invoked the `didClose` referenced within the `createDropinInstance` function was updated and handled correctly.

Here's a simplified example. The whole point is to create a `thing` if we're mounted or ignore it if we're unmounted.

```js
const [thing, setThing] = useState();

useEffect(() => {
    let unmounted = false;
    async function attach() {
        const ref = await createThingThatNeedsDom();

        // If `unmounted` is still "false" by the time the awaited function resolves we set the ref in state
        // but if it is "true" because we unmounted then we should just cleanup and never set state.
        if (unmounted) {
            ref.cleanup();
        }
        else {
            setThing(ref);
        }
    }

    // Run the async attach.
    attach();

    return () => { 
        unmounted = true;
    }
}, [
    // No dependencies because this component runs on mount.
]);
```

I've provided this example as a way to show you how it was used in the past because as implemented, `createDropinInstance` can break if we unmount before we get to `setDropinInstance(...)`. I realize you broke out the creation function because you need the logic twice but it doesn't make sense as written now. `createDropinInstance(didClose);` will always be called with `false`.

Please refactor this to correctly utilize the previous logic.